File: 0014-longest-common-prefix/0014-longest-common-prefix.cpp
class Solution {
private:
    string commonPrefix(string& left, string& right){
        int mn = min(left.size(), right.size());
        
        for (int i = 0; i < mn; ++i){
            if (left[i] != right[i])
                return left.substr(0, i);
        }
        
        return left.substr(0, mn);
    }
    
    string longestCommonPrefix(vector<string>& strs, int left, int right){
        if (left == right) return strs[left];
        else {
            int mid = left + ((right - left) >> 1);
            string lcpLeft = longestCommonPrefix(strs, left, mid);
            string lcpRight = longestCommonPrefix(strs, mid + 1, right);
            return commonPrefix(lcpLeft, lcpRight);
        }
    }
    
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty() || strs[0].size() == 0) return "";
        
        return longestCommonPrefix(strs, 0, strs.size() - 1);
    }
};

File: 0014-longest-common-prefix/0014-longest-common-prefix.py
class Solution:
    def commonPrefix(self, left: str, right: str) -> str:
        mn = min(len(left), len(right))
        for i in range(mn):
            if left[i] != right[i]:
                return left[:i]
        
        return left[:mn]
    
    def _longestCommonPrefix(self, s: List[str], l: int, r: int) -> str:
        if l == r:
            return s[l]
        else:
            mid = (l + r) >> 1
            lcp_left = self._longestCommonPrefix(s, l, mid)
            lcp_right = self._longestCommonPrefix(s, mid + 1, r)
            return self.commonPrefix(lcp_left, lcp_right)
    
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs or len(strs) == 0:
            return ""
        return self._longestCommonPrefix(strs, 0, len(strs) - 1)